File: linked-list/MyLinkedList.go
// go run MyLinkedList.go

package main

import (
    "fmt"
    "errors"
)

type Node[V any] struct {
    next *Node[V]
    value V
}

type LinkedList[T any] struct {
    head *Node[T]
}

// Go does not have classes.
// We can define methods on types though.
func (l *LinkedList[T]) addLast(value T) {
    newNode := Node[T]{nil, value}

    if l.head == nil {
        l.head = &newNode
    } else {
        current := l.head

        for current.next != nil {
            current = current.next
        }

        current.next = &newNode
    }
}

// * is a pointer receiver which allows mutation of type value
func (l *LinkedList[T]) addFirst(value T) {
    newNode := Node[T]{nil, value}

    if l.head == nil {
        l.head = &newNode
    } else {
        newNode.next = l.head
        l.head = &newNode
    }
}

// functions that begin with lower case are "private" and can
// only be used within the file, use capital letter to make it
// "exportable" to outside
func (l *LinkedList[T]) removeFirst() {
    if l.head != nil {
        if l.head.next == nil {
            l.head = nil
        } else {
            l.head = l.head.next
        }
    }
}

func (l *LinkedList[T]) removeLast() {
    if l.head != nil {
        current := l.head

        for current.next.next != nil {
            current = current.next
        }

        current.next = nil
    }
}

// we can omit * pointer here since nothing is getting modified
func (l LinkedList[T]) getFirst() (T, error) {
    if l.head != nil {
        return l.head.value, nil
    }

    var ret T

    return ret, errors.New("empty list");
}

// multiple return types are allowed
func (l LinkedList[T]) getLast() (T, error) {
    if l.head != nil {
        current := l.head

        for current.next != nil {
            current = current.next
        }

        return current.value, nil
    }

    var ret T

    return ret, errors.New("empty list");
}

func (l LinkedList[T]) display() {
    if l.head != nil {
        current := l.head

        for current != nil {
            fmt.Printf("[%v]->", current.value)

            current = current.next
        }
    }
}

func main() {
    // int linked list
    ll := LinkedList[int]{}

    ll.addLast(4)
    ll.addLast(3)
    ll.addLast(2)
    ll.addLast(1)
    ll.addFirst(5)
    ll.addFirst(6)
    ll.addFirst(7)

    ll.removeFirst()
    ll.removeLast()

    ll.display()

    first, err := ll.getFirst()

    if err == nil {
        fmt.Println("\nFirst Item: ", first)
    }

    second, err := ll.getLast()

    if err == nil {
        fmt.Println("Last Item: ", second)
    }

    // string linked list
    ll2 := LinkedList[string]{}

    ll2.addLast("tarantula")
    ll2.addLast("vulture")
    ll2.addLast("yak")
    ll2.addLast("zombat")
    ll2.addFirst("cat")
    ll2.addFirst("bat")
    ll2.addFirst("ant")

    ll2.removeFirst()
    ll2.removeLast()

    ll2.display()

    first2, err := ll2.getFirst()

    if err == nil {
        fmt.Println("\nFirst Item: ", first2)
    }

    second2, err := ll2.getLast()

    if err == nil {
        fmt.Println("Last Item: ", second2)
    }
}

File: linked-list/MyLinkedList.ts
// ts-node MyLinkedList.ts

class MyNode<T> { // Named this way to avoid global Node name
    next: MyNode<T> | null = null;
    value: T;

    constructor(val: T) {
      this.value = val;
    }
  }

  class LinkedList<T> {
    head: MyNode<T> | null = null;

    constructor() {
    }

    addFirst(val: T): void {
      const newNode = new MyNode<T>(val);

      if (this.head === null) {
        this.head = newNode;
      } else {
        newNode.next = this.head;

        this.head = newNode;
      }
    }

    addLast(val: T): void {
      const newNode = new MyNode<T>(val);

      if (this.head === null) {
        this.head = newNode;
      } else {
        let current = this.head;

        while (current !== null && current.next !== null) {
          current = current.next;
        }

        current.next = newNode;
      }
    }

    removeFirst(): void {
      if (this.head !== null) {
        if (this.head.next) {
          this.head = this.head.next;
        } else {
          this.head = null;
        }
      }
    }

    removeLast(): void {
      if (this.head !== null) {
        if (this.head.next === null) {
          this.head = null;
        } else {
          let current = this.head;

          if (current !== null) {
            while (current.next !== null && current.next.next !== null) {
              current = current.next;
            }

            current.next = null;
          }
        }
      }
    }

    getFirst(): T | never {
      if (this.head === null) {
        throw new Error("List is empty");
      }

      return this.head.value;
    }

    getLast(): T | never {
      if (this.head === null) {
        throw new Error("List is empty");
      }

      let current = this.head;

      while(current.next !== null) {
          current = current.next;
      }

      return current.value;
    }

    display(): void {
      let current = this.head;

      while (current) {
        console.log(`[${current.value}]->`);

        current = current.next;
      }
    }
  }

  // Number linkedlist
  const ll = new LinkedList<number>();

  ll.addLast(4);
  ll.addLast(3);
  ll.addLast(2);
  ll.addLast(1);
  ll.addFirst(5);
  ll.addFirst(6);
  ll.addFirst(7);

  ll.removeFirst();
  ll.removeLast();

  ll.display();

  try {
    console.log(`First Item: ${ll.getFirst()}`);
    console.log(`Last Item: ${ll.getLast()}`);
  } catch (e) {
    console.error("List is empty");
  }

  // String Linkedlist
  const ll2 = new LinkedList<string>();

  ll2.addLast("Cat");
  ll2.addLast("Dog");
  ll2.addFirst("Bat");
  ll2.addFirst("Ant");

  ll2.removeFirst();
  ll2.removeLast();

  ll2.display();

  try {
    console.log(`First Item: ${ll2.getFirst()}`);
    console.log(`Last Item: ${ll2.getLast()}`);
  } catch (e) {
    console.error("List is empty");
  }